pkg/conversion/mapi2capi/machine.go

```diff
@@ -0,0 +1,158 @@
+			// NodeDrainTimeout: ,
+			// NodeVolumeDetachTimeout: ,
+			// NodeDeletionTimeout: ,
+		},
+	}
+
+	// lifecycleHooks are handled via an annotation in Cluster API.
+	lifecycleAnnotations := getCAPILifecycleHookAnnotations(mapiMachine.Spec.LifecycleHooks)
+	for key, value := range lifecycleAnnotations {
+		capiMachine.Annotations[key] = value
+	}
+
+	errs = append(errs, setMAPINodeLabelsToCAPIManagedNodeLabels(field.NewPath("spec", "metadata", "labels"), mapiMachine.Spec.ObjectMeta.Labels, capiMachine.Labels)...)
+
+	// Unused fields - Below this line are fields not used from the MAPI Machine.
+
+	// mapiMachine.Spec.AuthoritativeAPI - Ignore as this is part of the conversion mechanism.
+
+	// metadata.labels - needs special handling
+	// metadata.annotations - needs special handling
+
+	errs = append(errs, handleUnsupportedMachineFields(mapiMachine.Spec)...)
+
+	return capiMachine, errs
+}
+
+func setMAPINodeLabelsToCAPIManagedNodeLabels(fldPath *field.Path, mapiNodeLabels map[string]string, capiNodeLabels map[string]string) field.ErrorList {
+	if len(mapiNodeLabels) == 0 {
+		return field.ErrorList{}
+	}
+
+	if capiNodeLabels == nil {
+		capiNodeLabels = map[string]string{}
+	}
+
+	errs := field.ErrorList{}
+
+	// TODO(OCPCLOUD-XXXX): Not all the labels on the CAPI Machine are propagated down to the corresponding CAPI Node, only the "CAPI Managed ones" are.
+	// These are those prefix by "node-role.kubernetes.io" or in the domains of "node-restriction.kubernetes.io" and "node.cluster.x-k8s.io".
+	// See: https://github.com/kubernetes-sigs/cluster-api/pull/7173
+	// and: https://github.com/fabriziopandini/cluster-api/blob/main/docs/proposals/20220927-label-sync-between-machine-and-nodes.md
+	for k, v := range mapiNodeLabels {
+		if !conversionutil.IsCAPIManagedLabel(k) {
+			errs = append(errs, field.Invalid(fldPath.Key(k), v, "label propogartion is not currently supported for this label"))
+		}
+
+		capiNodeLabels[k] = v
+	}
+
+	return errs
+}
+
+// getCAPILifecycleHookAnnotations returns the annotations that should be added to a CAPI Machine to represent the lifecycle hooks.
+func getCAPILifecycleHookAnnotations(hooks mapiv1.LifecycleHooks) map[string]string {
+	annotations := make(map[string]string)
+
+	for _, hook := range hooks.PreDrain {
+		annotations[fmt.Sprintf("%s/%s", capiv1.PreDrainDeleteHookAnnotationPrefix, hook.Name)] = hook.Owner
+	}
+
+	for _, hook := range hooks.PreTerminate {
+		annotations[fmt.Sprintf("%s/%s", capiv1.PreTerminateDeleteHookAnnotationPrefix, hook.Name)] = hook.Owner
+	}
+
+	return annotations
+}
+
+// handleUnsupportedMachineFields checks for fields that are not supported by CAPI and returns a list of errors.
+func handleUnsupportedMachineFields(spec mapiv1.MachineSpec) field.ErrorList {
+	var errs field.ErrorList
+
+	fldPath := field.NewPath("spec")
+
+	// ObjectMeta related fields should never get converted (aside from labels and annotations).
```

Comment:
Are callers expected to clear these fields before handing the struct to the `fromMAPIMachineToCAPIMachine` function?

Reply:
No, I'm not expecting users in their clusters to have these fields set, since they don't actually do anything if you set them (the controller ignores them).

So we return an error. They'll have to clear them if they want to proceed.